Add error handling/warnings for spreadsheet typos

airandfingers opened the pull request:
Fixes #30:

  - [x] Change `getAllServices` to use `getSheetTitles` instead of `index`
  - [x] Various additions to `api.js`:
    - [x] Allow overriding `SHEET_ID` in `.env`
    - [x] Define `getSheetTitles` utility using base `spreadsheets` endpoint
    - [x] Change `useAPI:error` to `errorMessage`, with custom message for mismatched `typeID`s
  - [x] Create `Error` component:
    - [x] Display `useAPI:errorMessage`, centered based on `Loader` styles
  - [x] Create `ReportLink` component ("Report this issue to the developers" button):
    - [x] Only shown if a `SUPPORT_EMAIL_ADDRESS` env variable is configured and custom error message is shown (mismatched `typeID`s)
    - [x] `mailto` link starts composing an email to `SUPPORT_EMAIL_ADDRESS` with subject and body set to custom error message
  - [x] Change all API-using pages (categories, types, services, and service-details) to render `Error` component instead of `<p>Something went wrong!</p>`
  - ~Add warnings for mismatched category fields; prevent types from being hidden due to mismatched category slugs~ (Removed due to comments; code still in https://github.com/airandfingers/fd-servicedirectory/tree/console-warnings)
  - ~Tests~ (Looks like the only tests are in `utils.test.js`, and I didn't modify `utils.js)

To test, add `SHEET_ID=1OC8f640o9zxr9WSOUk6rhTlC3eHuNqgFw_cEo7lBEhA` to `.env`; this points to a public copy of the Google Sheet with "pet-health" in the index instead of "pet-animal-health".

airandfingers commented:
@Alex-Cannon Sorry for the delay between commits!

See my PR description; I haven't merged master, written tests, or tested errors beyond mismatched `typeID`s, but I think I've handled those errors sufficiently well.

airandfingers commented:
I've merged master and handled category fields, and it doesn't look like we have tests yet other than `utils`, which I didn't touch.

Going to look at icons, after which I'll consider this good to go until/unless I hear otherwise from you.

Alex-Cannon commented:
@airandfingers great work so far. Once you're done, let @Vpr99 or I test and review your work, then we can merge. Thanks!

airandfingers commented on src/core/api.js:
```diff
@@ -3,16 +3,32 @@ import { useEffect, useReducer } from "react";
 
-// Utility for fetching a single sheet from a spreadsheet by its' title
+// Utility for fetching metadata about the sheets in the spreadsheet
+const getSheetTitles = async () => {
```
This new request gets the sheet titles all at once from the sheet's metadata, preventing `getAllServices` (and thus Search) from breaking if a type slug is listed wrong in the Index sheet.

Alex-Cannon replied:
This sounds like a good change. Awesome.

airandfingers commented on src/core/api.js:
```diff
@@ -61,9 +77,28 @@ export const useAPI = (fetcher, opts) => {
+              status === "INVALID_ARGUMENT" &&
+              message.startsWith("Unable to parse range:")
+            ) {
+              errorMessage = `Failed to get services for type ${opts}`;
```
This custom error message (and a report button, if an email is configured) will display on the services page (e.g. ``/categories/medical/pet-health`) if the slug doesn't match a sheet title.

For any other errors, the default message "Something went wrong!" will still be displayed, with no report button.

Alex-Cannon replied:
Like I was saying above, a better way to do this would be to put red text on the home `/` page with something like: "There were X number of errors processing your data. Site admins click here to view a full log."

This way site admins don't have to click every category in order to check for errors.

airandfingers commented on src/pages/categories/index.jsx:
```diff
@@ -1,10 +1,11 @@
-import { uniqBy } from "lodash";
```
This was the only thing imported from `lodash`, so we should consider removing the dependency on it.

Alex-Cannon replied:
I agree with you, but let's focus on error handling/warnings in the PR. I'll create a new issue for this.

airandfingers commented on src/pages/categories/index.jsx:
```diff
@@ -23,26 +24,46 @@ const IntroText = styled(H1)({
+const getCategories = data => {
+  const categoriesBySlug = {};
+  data.forEach(([title, slug, description, icon, color]) => {
+    const existingCategory = categoriesBySlug[slug];
```
Previously, duplicates were eliminated based on `title`, so if two categories had the same title but different slugs, then one wouldn't show up on the types page (e.g. `/categories/medical`).

With this change, in the above case we'd see duplicate categories with the same title, each linking to a separate types page that only lists categories with that category slug.

Alex-Cannon replied:
Good idea, this is a great change. :)

airandfingers commented on src/pages/categories/index.jsx:
```diff
@@ -52,14 +73,7 @@ const Categories = () => {
-              link={c.slug}
-              title={c.title}
-            />
+            <CategoryCard key={c.slug} {...c} />
```
This syntax is cleaner IMO whenever you're passing 4+ props like `propName={propName}`.

This `key={c.slug}` should be avoidable by adding a `key` property to `category` above, but ESLint refused to let me commit that, since it couldn't detect the `key` prop.

Alex-Cannon replied:
Good change

airandfingers commented on src/pages/service-detail/index.jsx:
```diff
@@ -57,7 +56,10 @@ const ServiceDetail = ({ match }) => {
+        backLink={
+          new URLSearchParams(location.search).get("backLink") ||
+          `/categories/${categoryId}/${typeId}`
+        }
```
ESLint made this whitespace change automatically.

airandfingers commented on src/pages/service-detail/index.jsx:
```diff
@@ -29,19 +30,17 @@ const PhoneLink = styled.a({
   textDecoration: "none",
 });
 
-const urlQuery = new URLSearchParams(location.search);
```
ESLint woudn't let me commit while this unused variable was declared.

airandfingers commented:
@Alex-Cannon @Vpr99 done, take a look when you get the chance!

I added a "review" with some annotations for your reviewing convenience.

Alex-Cannon commented on src/components/error/index.js:
```diff
@@ -0,0 +1,41 @@
+import { P1 } from "~/components/typography";
+import Button from "~/components/button";
+
+const reportEmail = process.env.SUPPORT_EMAIL_ADDRESS;
```
Definitely a great idea.

Alex-Cannon commented on .eslintrc.js:
```diff
@@ -14,6 +14,7 @@ module.exports = {
   parser: 'babel-eslint',
   plugins: ['jsx-a11y', 'react', 'prettier', 'react-hooks'],
   rules: {
+    'no-console': 'off',
```
Hmm. I understand what you're saying, but I'd consult @Vpr99 Before reverting this. @Vpr99 setup the app and there was probably a reason for this.

What people at the fire department will look at the console? I'm not sure anyone would know how to even open it. Outlawing `console` messages makes our code cleaner as well as the log cleaner. I think we can implement this feature and keep our code cleaner.

A better solution might be to have text somewhere saying "There were x number of errors processing non-profit data. Admins please click here to view a full log," that way it's a lot more user-friendly for them.

I'd be happy to implement this or you could if you like.

Alex-Cannon commented on src/core/api.js:
```diff
@@ -61,9 +77,28 @@ export const useAPI = (fetcher, opts) => {
           dispatch({ type: "SUCCESS", payload: data });
         }
       } catch (e) {
+        console.error(
```
I don't think people at the fire department would open the browser console to view errors, so we should create an error page for admins like I mentioned in other parts of my review. 

The goal with adding this PR is to allow the Fire Department to maintain the spreadsheet without any help. Having an error page is a lot easier for them than messing with the console.

I'd be happy to implement this or you could if you like. Either way, we gotta do something other than use the console before we close this issue.

airandfingers replied:
Gotcha, I didn't realize the point was to enable PFD to maintain the spreadsheet themselves, but that makes sense.

It'll get a bit tricky to implement such an error page (and per-page "There were X errors" messages) with the current structure; I imagine we'd want global state and a global component like `Nav`.

As I suggested in https://github.com/CodeForFoco/fd-servicedirectory/issues/40, IMO ideally all data would be fetched once when the site is loaded, rather than per-page as needed. If we did that, we could validate all data at once, and we'd already have the global state to store any errors and display them on the error page.

airandfingers replied:
I should mention that I'd be willing/happy to switch the site to load all data at once, and that I'd try to find a way to get it to load whatever data is needed by the initial page first (to avoid making the initial load slower). I'd probably want to use the built-in Context API - a top-level `Provider` with per-component `useContext` calls.

I'm less excited about the prospect of implementing the error page and per-page messages under the current data-loading scheme, in part because of the many paths the code can take to load and validate the data (at least five: categories, types, services, service-details, search).

Alex-Cannon replied:
@airandfingers I think what we can do is this:

1. Take all your validation code and error class code and commit it. In `/utils`, for example, getCategories should be part of the api and your validation should be part of `/utils` that getCategories uses.

2. We merge your code, but keep the issue open

3. We create a new issue to implement global state

4. Once global state is implemented, we can create these issues to allow error logging:
    1. Implement sheet error catching
    2. Create Error Page
    3. Create Error Component

Alex-Cannon commented on src/pages/categories/index.jsx:
```diff
@@ -23,26 +24,46 @@ const IntroText = styled(H1)({
+    if (theme.colors[color] === undefined) {
+      console.warn(`Unknown icon color: ${color}`);
+    }
+    if (!existingCategory) {
```
We should push this to the top of the function and have something like so (with categoryExists being a util function):

```javascript
if (!categoryExists(slug, data)) return;

categoriesBySlug[slug] = {
  color,
  description
  ...
}
```

Which is less verbose and more clear than defining `category`, `existingCategory`, and having an `if` statement later on. Would you make this change please?

airandfingers replied:
Hm, the problem is that I'm not just returning when a category exists already, I'm comparing `existingCategory` and `category` to warn about any values that differ between them. Only `existingCategory` will get returned by this function, so if we want to detect discrepancies, we have to do it here.

airandfingers commented:
> Also, @airandfingers would you provide more details on how to test this? I followed your instructions and I'm not sure where I should be looking for these error messages or what errors should be showing up.

Sure, so you've added `SHEET_ID=1OC8f640o9zxr9WSOUk6rhTlC3eHuNqgFw_cEo7lBEhA` to `.env`; you should see the following behavior:

1. G11 says "pet-animal-health1" instead of "pet-animal-health":
  a. Clicking Medical > Pet Health should take you to `/categories/medical/pet-animal-health1`, where you should see "Failed to get services for type pet-animal-health1" and the "Report this issue to the developers" button if you have an `SUPPORT_EMAIL_ADDRESS` in your `.env`.
  b. The search page should still work as designed (previously it showed "Something went wrong!"
2. B2 says "basic-needs1" instead of "basic-needs" - on `/categories` you should see two "Basic Needs" entries, one linking to `/categories/basic-needs1` and showing only the type "Food"; the other linking to `/categories/basic-needs` and showing the rest of the types.
3. D7 says "conversation1" instead of "conversation" - on `/categories` the "Mental Health/Addiction" entry shouldn't show an icon, and you should see a console warning about an unknown icon "conversation1".
4. E9 says "fire1" instead of "fire" - on `/categories` the "Medical" entry shows up as black on white instead of red on light red, and you should see a console warning about an unknown icon color "fire1".
5. A8 says "Mental Health/Addiction1" instead of "Mental Health/Addiction" - on `categories` you should see a console warning about the different category titles.
6. C10 says "Medical Services and Equipment1" instead of "Medical Services and Equipment" - on `/categories` you should see a console warning about the different category descriptions.

Note: the differences between the effects of cases 3-4 (rendered changes and warning) and 5-6 (just warning) is that 3 and 4 are mistakes in the first entry for a category, which is returned by `getCategories`, while 5 and 6 are mistakes in later entries, which are not returned by it.

Alex-Cannon commented on .eslintrc.js:
```diff
@@ -14,6 +14,7 @@ module.exports = {
   parser: 'babel-eslint',
   plugins: ['jsx-a11y', 'react', 'prettier', 'react-hooks'],
   rules: {
+    'no-console': 'off',
```
Don't log to the console, admins won't look at it. Would you revert this? Thanks!

Alex-Cannon commented on src/components/error/index.js:
```diff
@@ -0,0 +1,41 @@
+      <ReportLink
+        target="_blank"
+        href={`mailto:${reportEmail}?subject=${errorMessage}&body=${errorMessage}`}
+      >
```
An error component is a great idea. It's good that this replaces ```<p> Something went wrong! </p>``` in the app. Instead of emailing developers, this should link to a `/errors` page where admins can view their error log.

Would you...

1. Remove the ReportLink?

We can create an error page later. Thanks!

Alex-Cannon commented on src/components/icon/icons.jsx:
```diff
@@ -58,6 +58,7 @@ const Icons = ({ icon }) => {
         />
       );
     default:
+      console.warn("Unknown icon type:", icon);
```
Admins won't look at the log, so you can go ahead and remove this. Thanks!

Alex-Cannon commented on src/core/api.js:
```diff
@@ -61,9 +77,28 @@ export const useAPI = (fetcher, opts) => {
           dispatch({ type: "SUCCESS", payload: data });
         }
       } catch (e) {
+        console.error(
```
We want to remove console logging as admins won't look at the logs.

Alex-Cannon commented on src/pages/categories/index.jsx:
```diff
@@ -23,26 +24,46 @@ const IntroText = styled(H1)({
+// Transforms the data into usable objects and filter out duplicates
+const getCategories = data => {
+  const categoriesBySlug = {};
+  data.forEach(([title, slug, description, icon, color]) => {
```
We can revert most of this function as we aren't worrying about console logging. I do think it'd be a good idea to collect a list of errors here in the future (don't have to worry about it for this PR) that we can put on the `/errors` page.

This is my bad for not being clear as to why we're handling errors.

airandfingers commented:
Okay @Alex-Cannon, I made the changes you requested (or at least I think I did), please re-test and re-review.

I updated the tasks in the PR description, including a link to https://github.com/airandfingers/fd-servicedirectory/tree/console-warnings, which contains the `console` warnings, to be referenced whenever we implement errors in their place.

Updated test instructions after adding `SHEET_ID=1OC8f640o9zxr9WSOUk6rhTlC3eHuNqgFw_cEo7lBEhA` to `.env`:

1. G11 says "pet-animal-health1" instead of "pet-animal-health":
  a. Clicking Medical > Pet Health should take you to `/categories/medical/pet-animal-health1`, where you should see "Failed to get services for type pet-animal-health1".
  b. The search page should still work as designed (previously it showed "Something went wrong!")
2. B2 says "basic-needs1" instead of "basic-needs" - on `/categories` you should see two "Basic Needs" entries, one linking to `/categories/basic-needs1` and showing only the type "Food"; the other linking to `/categories/basic-needs` and showing the rest of the types.
3. D7 says "conversation1" instead of "conversation" - on `/categories` the "Mental Health/Addiction" entry shouldn't show an icon (no change from before).
4. E9 says "fire1" instead of "fire" - on `/categories` the "Medical" entry shows up as black on white instead of red on light red (no change from before).
5. A8 says "Mental Health/Addiction1" instead of "Mental Health/Addiction" - ignored (no change from before).
6. C10 says "Medical Services and Equipment1" instead of "Medical Services and Equipment" - ignored (no change from before).
